File: tools/rosrecord/src/rosrecord/rosrecord.cpp
/*********************************************************************
* Software License Agreement (BSD License)
*
*  Copyright (c) 2008, Willow Garage, Inc.
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*   * Neither the name of Willow Garage, Inc. nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
********************************************************************/

#include <time.h>
#include <sys/stat.h>
#include "ros/ros.h"
#include "rosrecord/Recorder.h"
#include "topic_tools/shape_shifter.h"

#include <boost/thread.hpp>

#include <sstream>
#include <string>
#include <set>
#include <queue>

#include "std_msgs/Empty.h"

class OutgoingMessage
{
public:
  OutgoingMessage(std::string _topic_name, topic_tools::ShapeShifter::ConstPtr _msg, ros::Time _time) :
    topic_name(_topic_name), msg(_msg), time(_time) {}

  std::string topic_name;
  topic_tools::ShapeShifter::ConstPtr msg;
  ros::Time time;
};

class OutgoingQueue
{
public:
  OutgoingQueue(std::string _fname, std::queue<OutgoingMessage>* _queue, ros::Time _time) :
    fname(_fname), queue(_queue), time(_time) {}

  std::string fname;
  std::queue<OutgoingMessage>* queue;
  ros::Time time;
};


//! Global verbose flag so we can easily use from callback
bool g_verbose = false;

//! Global snapshot flag so we can easily use from callback
bool g_snapshot = false;

// Global Eventual exit code
int  g_exit_code       = 0;

//! Global variable including the set of currenly recording topics
std::set<std::string> g_currently_recording;

//! Global variable used for initialization of counting messages
int g_count = -1;

//! Global variable used for book-keeping of our number of subscribers
int g_num_subscribers = 0;

//! Global queue for storing 
std::queue<OutgoingMessage>* g_queue;

//! Global queue size
uint64_t g_queue_size=0;

//! Global max queue size
uint64_t g_max_queue_size=1048576*256;

//! Queue of queues to be used by the snapshot recorders
std::queue<OutgoingQueue> g_queue_queue;

//! Mutex for global queue
boost::mutex g_queue_mutex;

//! Conditional variable for global queue
boost::condition_variable_any g_queue_condition;

//! Compression mode
static enum {
  COMPRESSION_NONE,
  COMPRESSION_GZIP,
  COMPRESSION_BZIP2
} g_compression = COMPRESSION_NONE;

template <class T>
std::string time_to_str(T ros_t)
{
  char buf[1024]      = "";
  time_t t = ros_t.sec;
  struct tm *tms = localtime(&t);
  strftime(buf, 1024, "%Y-%m-%d-%H-%M-%S", tms);
  return std::string(buf);
}

//! Helper function to print executable usage
void print_usage() {
  fprintf (stderr, "Usage: rosrecord [options] TOPIC1 [TOPIC2 TOPIC3...]\n"
                   "  rosrecord logs ROS message data to a file.\n"
                   );
}


//! Helper function to print executable options
void print_help() {
  print_usage();
  fprintf(stderr, "Options:\n");
  fprintf(stderr, " -c <num>    : Only receive <num> messages on each topic\n");
  fprintf(stderr, " -f <prefix> : Prepend file prefix to beginning of bag name (name will always end with date stamp)\n");
  fprintf(stderr, " -F <fname>  : Record to a file named exactly <fname>.bag\n");
  fprintf(stderr, " -a          : Record all published messages.\n");
  fprintf(stderr, " -v          : Display a message every time a message is received on a topic\n");
  fprintf(stderr, " -m          : Maximize internal buffer size in MB (Default: 256MB)  0 = infinite.\n");
  fprintf(stderr, " -z          : Compress the messages with gzip\n");
  fprintf(stderr, " -j          : Compress the messages with bzip2\n");
  fprintf(stderr, " -s          : (EXPERIMENTAL) Enable snapshot recording (don't write to file unless triggered)\n");
  fprintf(stderr, " -t          : (EXPERIMENTAL) Trigger snapshot recording\n");
  fprintf(stderr, " -h          : Display this help message\n");
}


//! Callback to be invoked to save messages into a queue
void do_queue(topic_tools::ShapeShifter::ConstPtr msg,
               std::string topic_name,
               boost::shared_ptr<ros::Subscriber> subscriber,
               boost::shared_ptr<int> count)
{
  ros::Time rectime = ros::Time::now();

  if (g_verbose)
    std::cout << "Received message on topic " << subscriber->getTopic() << std::endl;

  OutgoingMessage out(topic_name, msg, rectime);

  {
    boost::mutex::scoped_lock lock(g_queue_mutex);
    g_queue->push(out);
    g_queue_size += out.msg->msgBufUsed;

    while (g_max_queue_size > 0 && g_queue_size > g_max_queue_size)
    {
      OutgoingMessage drop = g_queue->front();
      g_queue->pop();
      g_queue_size -= drop.msg->msgBufUsed;
      if (!g_snapshot)
      {
        static ros::Time last = ros::Time();
        ros::Time now = ros::Time::now();
        if (now > last + ros::Duration(5.0))
        {
          ROS_WARN("Rosrecord buffer exceeded.  Dropping oldest queued message.");
          last = now;
        }
      }
    }
  }
  
  if (!g_snapshot)
    g_queue_condition.notify_all();

  // If we are book-keeping count, decrement and possibly shutdown
  if ((*count) > 0)
  {
    (*count)--;
    if ((*count) == 0)
    {
      subscriber->shutdown();
      
      if (--g_num_subscribers == 0)
        ros::shutdown();
    }
  }
}

//! Callback to be invoked to actually do the recording
void snapshot_trigger(std_msgs::Empty::ConstPtr trigger, std::string prefix, bool add_date)
{
  ros::WallTime rectime = ros::WallTime::now();

  std::vector<std::string> join;

  if (prefix.length() > 0)
    join.push_back(prefix);
  if (add_date)
    join.push_back(time_to_str(rectime));

  std::string tgt_fname = join[0];
  for (size_t i = 1; i < join.size(); i++)
    tgt_fname = tgt_fname + "_" + join[i];

  tgt_fname = tgt_fname + std::string(".bag");


  ROS_INFO("Triggered snapshot recording with name %s.", tgt_fname.c_str());

  {
    boost::mutex::scoped_lock lock(g_queue_mutex);

    OutgoingQueue out(tgt_fname, g_queue, ros::Time::now());

    g_queue = new std::queue<OutgoingMessage>;
    g_queue_size=0;

    g_queue_queue.push(out);
  }
  g_queue_condition.notify_all();
}

//! Thread that actually does writing to file.
void do_record(std::string prefix, bool add_date)
{
  ros::WallTime rectime = ros::WallTime::now();

  std::vector<std::string> join;

  if (prefix.length() > 0)
    join.push_back(prefix);
  if (add_date)
    join.push_back(time_to_str(rectime));

  std::string tgt_fname = join[0];
  for (size_t i = 1; i < join.size(); i++)
    tgt_fname = tgt_fname + "_" + join[i];

  std::string suffix;
  switch (g_compression) {
  case COMPRESSION_GZIP:
    suffix = ".gz";
    break;
  case COMPRESSION_BZIP2:
    suffix = ".bz2";
    break;
  case COMPRESSION_NONE:
    break;
  default:
    ROS_FATAL("Unknown compression method requested: %d", g_compression);
    g_exit_code = 1;
    ros::shutdown();
    break;
  }

  tgt_fname = tgt_fname + std::string(".bag") + suffix;
  std::string fname = tgt_fname + std::string(".active") + suffix;

  ROS_INFO("Recording to %s.", tgt_fname.c_str());

  ros::NodeHandle nh;
  ros::record::Recorder recorder;

  // Open our recorder and add available topics
  if (!recorder.open(std::string(fname)))
  {
    ROS_FATAL("Could not open output file: %s", fname.c_str());
    g_exit_code = 1;
    ros::shutdown();
  }
  
  recorder.writeVersion();
  recorder.writeFileHeader();

  // Technically the g_queue_mutex should be locked while checking empty
  // Except it should only get checked if the node is not ok, and thus
  // it shouldn't be in contention.
  while (nh.ok() || !g_queue->empty())
  {
    boost::unique_lock<boost::mutex> lock(g_queue_mutex);

    bool finished = false;
    while(g_queue->empty())
    {
      if (!nh.ok())
      {
        lock.release()->unlock();
        finished = true;
        break;
      }
      g_queue_condition.wait(lock);
    }
    if (finished)
      break;

    OutgoingMessage out = g_queue->front();
    g_queue->pop();
    g_queue_size -= out.msg->msgBufUsed;

    lock.release()->unlock();

    recorder.record(out.topic_name, out.msg, out.time);
  }

  // Write the index to a file
  recorder.writeIndex();

  // Close the file nicely
  recorder.close();
  rename(fname.c_str(),tgt_fname.c_str());
}

void do_record_bb()
{
  ros::NodeHandle nh;
  
  while (nh.ok() || !g_queue_queue.empty())
  {
    boost::unique_lock<boost::mutex> lock(g_queue_mutex);
    while(g_queue_queue.empty())
    {
      if (!nh.ok())
        return;
      g_queue_condition.wait(lock);
    }

    OutgoingQueue out_queue = g_queue_queue.front();
    g_queue_queue.pop();

    lock.release()->unlock();

    std::string tgt_fname = out_queue.fname;
    std::string fname = tgt_fname + std::string(".active");

    ros::record::Recorder recorder;

    if (recorder.open(fname))
    {
      recorder.writeVersion();

      while (!out_queue.queue->empty())
      {
        OutgoingMessage out = out_queue.queue->front();
        out_queue.queue->pop();
        recorder.record(out.topic_name, out.msg, out.time);
      }

      // Close the file nicely
      recorder.close();

      // Rename the file to the actual target name
      rename(fname.c_str(),tgt_fname.c_str());
    } else {
      ROS_ERROR("Could not open file: %s", out_queue.fname.c_str());
    }
  }
}

void do_check_master(const ros::TimerEvent& e, ros::NodeHandle& node_handle)
{
  ros::master::V_TopicInfo all_topics;

  if (ros::master::getTopics(all_topics))
  {
    // For each topic
    for (ros::master::V_TopicInfo::iterator topic_iter = all_topics.begin();
         topic_iter != all_topics.end(); 
         topic_iter++)
    {
      // If we're not currently recording it, do so
      if (g_currently_recording.find(topic_iter->name) == g_currently_recording.end())
      {
        boost::shared_ptr<int> count(new int(g_count));
        boost::shared_ptr<ros::Subscriber> sub(new ros::Subscriber);
        *sub = node_handle.subscribe<topic_tools::ShapeShifter>(topic_iter->name, 100, boost::bind(&do_queue, _1, topic_iter->name, sub, count));
        g_currently_recording.insert(topic_iter->name);
      }
    }
  }
}

void do_trigger()
{
  // Get a node_handle
  ros::NodeHandle node_handle;
  ros::Publisher pub = node_handle.advertise<std_msgs::Empty>("snapshot_trigger", 1, true);
  pub.publish(std_msgs::Empty());
  ros::Timer terminate_timer = node_handle.createTimer(ros::Duration(1.0), boost::bind(&ros::shutdown));
  ros::spin();  
}

//! Main function
int main(int argc, char **argv)
{
  ros::init(argc, argv, "rosrecord", ros::init_options::AnonymousName);

  // Variables
  bool check_master    = false; // Whether master should be checked periodically

  bool add_date = true;;
  std::string prefix("");   // Prefix                                          
  
  // Parse options  
  int option_char;

  while ((option_char = getopt(argc,argv,"f:F:c:m:ajzsthv")) != -1)
  {
    switch (option_char)
    {
    case 'f': prefix = std::string(optarg); break;
    case 'F': prefix = std::string(optarg); add_date = false; break;
    case 'c': g_count = atoi(optarg); break;
    case 'a': check_master = true; break;
    case 's': g_snapshot = true; break;
    case 't': do_trigger(); return 0; break;
    case 'v': g_verbose = true; break;
    case 'z': g_compression = COMPRESSION_GZIP; break;
    case 'j': g_compression = COMPRESSION_BZIP2; break;
    case 'm': 
      {
        int m=0;
        m=atoi(optarg);
        if (m < 0)
        {
          fprintf(stderr, "Buffer size must be 0 or positive.\n");
          return 1;
        }
        g_max_queue_size =  1048576*m;
      }
      break;
    case 'h': print_help(); return 1;
    case '?': print_usage(); return 1;
    }
  }

  if (g_snapshot)
    ROS_WARN("Using snapshot mode in rosrecord is experimental and usage syntax is subject to change");

  // Logic to make sure count is not specified with automatic topic
  // subscription (implied by no listed topics)
  if ((argc - optind) < 1)
  {
    if (g_count > 0)
    {
      fprintf(stderr, "Specifing a count is not valid with automatic topic subscription.\n");
      return 1;
    }
    if (!check_master)
    {
      ROS_WARN("Running rosrecord with no arguments has been deprecated.  Please use 'rosrecord -a' instead\n");
      check_master = true;
    }
  }

  g_queue = new std::queue<OutgoingMessage>;

  // Get a node_handle
  ros::NodeHandle node_handle;

  // Only set up recording if we actually got a useful nodehandle.
  if (node_handle.ok())
  {
    boost::thread record_thread;

    // Spin up a thread for actually writing to file
    if (!g_snapshot)
      record_thread = boost::thread(boost::bind(&do_record, prefix, add_date));
    else
      record_thread = boost::thread(boost::bind(&do_record_bb));

    ros::Subscriber trigger = node_handle.subscribe<std_msgs::Empty>("snapshot_trigger", 100, boost::bind(&snapshot_trigger, _1, prefix, add_date));

    // Every non-processed argument is assumed to be a topic
    for (int i = optind; i < argc; i++)
    {
      boost::shared_ptr<int> count(new int(g_count));
      boost::shared_ptr<ros::Subscriber> sub(new ros::Subscriber);
      g_num_subscribers++;
      *sub = node_handle.subscribe<topic_tools::ShapeShifter>(argv[i], 100, boost::bind(&do_queue, _1, argv[i], sub, count));
    }

    ros::Timer check_master_timer;
    if (check_master)
      check_master_timer = node_handle.createTimer(ros::Duration(1.0), boost::bind(&do_check_master, _1, boost::ref(node_handle)));
    
    ros::MultiThreadedSpinner s(10);
    ros::spin(s);
    
    g_queue_condition.notify_all();
    
    record_thread.join();
        
  }
  
  delete g_queue;

  // Return our exit code
  return g_exit_code;
}

File: tools/rosrecord/src/rosrecord.py
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Willow Garage, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

PKG = 'rosrecord'
import roslib; roslib.load_manifest(PKG)

import warnings
warnings.warn('rosrecord is deprecated; use the rosbag package (http://www.ros.org/wiki/rosbag) instead', category=DeprecationWarning)

import optparse
import os
import sys

import rospy
import rosbag
from bisect import bisect

class ROSRecordException(Exception):
    """
    Base exception type for rosrecord-related errors.
    """
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)

class BagReader(object):
    """
    Bag file reader
    """
    def __init__(self, f):
        """
        Open a bag file for reading
        
        @param f: either a path to bag file to open for reading, or a stream to read from 
        @type  f: str or file
        @raise ROSRecordException: if version not supported
        """
        try:
            self._bag = rosbag.Bag(f, 'r')
        except Exception, e:
            raise ROSRecordException(str(e))

    def close(self):
        """
        Close the bag file
        """
        try:
            if self._bag:
                self._bag.close()
        except Exception, e:
            raise ROSRecordException(str(e))

    @property
    def datatypes(self): return dict([(info.topic, info.datatype) for info in self._bag._connections.values()])
    
    @property
    def filename(self): return self._bag.filename

    def read_index(self):
        """
        Read the index from the file (if it exists)
    
        @return dict(topic, [(stamp, pos)...]): bag file index
        """
        try:
            index = {}
            
            connection_index = self._bag.get_index()
            for (connection_id, connection_topic, connection_datatype), entries in connection_index.items():
                if connection_topic not in index:
                    index[connection_topic] = []
                topic_index = index[connection_topic]

                for entry in entries:
                    topic_index.insert(bisect(topic_index, entry), entry)
            
            self.index = self._bag.get_index()

        except rosbag.ROSBagException:
            self.index = None

        return self.index

    def logplayer(self, raw=False, seek=None):
        if not self._bag:
            return

        if seek is not None:
            raise ROSRecordException('seek not supported')

        try:
            try:
                for topic, msg, t in self._bag.read_messages(raw=raw):
                    if msg is None or rospy.is_shutdown():
                        break
                    yield topic, msg, t
            except KeyboardInterrupt:
                pass # break iterator
        finally:
            self._bag.close()

    def raw_messages(self):
        """
        Generates raw messages
    
        @return (pos, topic, msg, t)...: tuple of position, topic, message data and time stamp
        """
        try:
            for topic, msg, t in self._bag.read_messages(raw=True):
                if msg is None or rospy.is_shutdown():
                    break
                (datatype, data, md5sum, position, pytype) = msg
                yield position, topic, msg, t
        
        except rosbag.ROSBagException, e:
            rospy.logerr('ROSBagException: couldn\'t read message - %s' % str(e))
        except IOError:
            rospy.logerr('IOError: couldn\'t read message')
        except KeyError:
            rospy.logerr('KeyError: couldn\'t read message')

    def load_message(self, pos, index=None):
        """
        Load message from the bag file at a given position
    
        @param pos: byte offset in file of start of message record
        @type  pos: int
        @return (datatype, msg, t): tuple of datatype, deserialized message, and timestamp   
        """
        msgs = list(self.load_messages([pos], index))
        if len(msgs) == 0:
            return None, None, None
        
        return msgs[0]
    
    ## Returns iterator of (datatype, message, timestamp)
    def load_messages(self, positions, index=None):
        try:
            for pos in positions:
                _, raw_msg, t = self._bag._read_message(pos, True)
                if raw_msg is not None:
                    (datatype, message_data, md5, bag_pos, pytype) = raw_msg
                    
                    msg = pytype()
                    msg.deserialize(message_data)

                    yield (datatype, msg, t)
        
        except rosbag.ROSBagException, e:
            rospy.logerr('ROSRecordException: couldn\'t read %s - %s' % (str(pos), str(e))) 
        except IOError:
            rospy.logerr('IOError: couldn\'t read %d' % pos)
        except KeyError:
            rospy.logerr('KeyError: couldn\'t read %d' % pos)

def logplayer(f, raw=False, seek=None):
    """
    Iterator for messages in a bag file. Return type of iterator depends on value of 'raw' parameter, which
    is provided for improved iteration performance.
    
    If raw is False:
      - topic: topic name (str)
      - msg: Message instance (Message)
      - t: time (rospy.Time)
    
     If raw is True:
      - topic: topic name (str)
      - msg: (datatype, data, md5sum, bag_position).
         - datatype: msg type name
         - data: serialized message bytes.
         - bag_position: offset into the file, which can be used for seeking.
         - md5sum: md5sum of msg type for identify msg version
         - NOTE: More elements may be added to this in the future, so this should
           be unpacked assuming non-fixed length.
      - t: time (rospy.Time)
     
    @param filename: name of file to playback from
    @type  filename: str
    @param raw: see description for raw behavior
    @type  raw: bool
    @param seek: initial position to seek to (default None)
    @type  seek: int
    @raise ROSRecordException: if file format is unrecognized
    @return: iterator for messages. See description for iterator behavior.
    @rtype: iterator
    """
    r = BagReader(f)
    
    return r.logplayer(raw, seek)

class Rebagger(object):
    """
    Utility class for writing Message instances to a ROS .bag file
    """
    def __init__(self, filename):
        self._bag = rosbag.Bag(filename, 'w')
    
    def add(self, topic, msg, t=None, raw=False):
        """
        Add a message to the bag
        @param msg: message to add to bag
        @type  msg: Message
        @param raw: if True, msg is in raw format (msg_type, serialized_bytes, md5sum, pytype)
        @param raw: bool
        @param t: ROS time of message publication
        @type  t: U{roslib.message.Time}
        """
        try:
            self._bag.write(topic, msg, t, raw)
        except:
            raise ROSRecordException('error adding message to bag')
    
    def close(self):
        try:
            if self._bag:
                self._bag.close()
        except:
            raise ROSRecordException('error closing Rebagger')

def rebag(inbag, outbag, filter_fn, verbose_pattern=None, raw=False):
    """
    Filter the contents of inbag to outbag using filter_fn
    @param inbag: filename of input bag file.
    @type  inbag: str
    @param outbag: filename of output bag file. Existing bag file will be overwritten
    @type  outbag: str
    @param filter_fn: Python function that returns True if msg is to be kept
    @type  filter_fn: fn(topic, msg, time)
    @param raw: if True, msg will be kept deserialized. This is useful
    if you are removing messages that no longer exist.
    @type  raw: bool
    @param verbose_pattern: Python function that returns string to
    print for verbose debugging
    @type  verbose_pattern: fn(topic, msg, time)
    """
    rebag = Rebagger(outbag)
    if verbose_pattern:
        for topic, msg, t in logplayer(inbag, raw=raw):
            if filter_fn(topic, msg, t):
                print "MATCH", verbose_pattern(topic, msg, t)
                rebag.add(topic, msg, t, raw=raw)          
                if rospy.is_shutdown():
                    break
            else:
                print "NO MATCH", verbose_pattern(topic, msg, t)          
    else: #streamlined
        for topic, msg, t in logplayer(inbag, raw=raw):
            if filter_fn(topic, msg, t):
                rebag.add(topic, msg, t, raw=raw)
                if rospy.is_shutdown():
                    break
    rebag.close()

def _rebag_main():
    """
    main routine for rosrebag command-line tool
    """

    rospy.logwarn("Rosrebag has been deprecated.  Please use 'rosbag filter' instead\n")

    ## filter function that uses command line expression to test topic/message/time
    def expr_eval(expr):
        def eval_fn(topic, m, t):
            return eval(expr)
        return eval_fn
    
    parser = optparse.OptionParser(usage="""Usage: %prog in.bag out.bag filter-expression

filter-expression can be any Python-legal expression.
The following variables are available:
 * topic: name of topic
 * m: message
 * t: time of message (t.secs, t.nsecs)
""", prog='rosrebag')
    parser.add_option('--print', dest="verbose_pattern", default=None,
                      metavar="PRINT-EXPRESSION", help="Python expression to print for verbose debugging. Uses same variables as filter-expression.")

    options, args = parser.parse_args()
    if len(args) == 0:
        parser.print_usage()
        sys.exit(0)
    elif len(args) != 3:
        parser.error("invalid arguments")

    inbag, outbag, expr = args
    if options.verbose_pattern:
        verbose_pattern = expr_eval(options.verbose_pattern)
    else:
        verbose_pattern = None    
    if not os.path.isfile(inbag):
        print >> sys.stderr, "cannot locate input bag file [%s]" % inbag
        sys.exit(1)
        
    rebag(inbag, outbag, expr_eval(expr), verbose_pattern=verbose_pattern)
